tests/unit/s2n_async_pkey_test.c

```diff
@@ -63,6 +67,59 @@ static int async_handler_wipe_connection_and_apply(struct s2n_connection *conn)
     return S2N_FAILURE;
 }
 
+static int async_handler_sign_with_different_pkey_and_apply(struct s2n_connection *conn)
+{
+    /* Check that we have pkey_op */
+    EXPECT_NOT_NULL(pkey_op);
+
+    /* Extract pkey */
+    struct s2n_cert_chain_and_key *chain_and_key = s2n_connection_get_selected_cert(conn);
+    EXPECT_NOT_NULL(chain_and_key);
+    s2n_cert_private_key *pkey = s2n_cert_chain_and_key_get_private_key(chain_and_key);
+    EXPECT_NOT_NULL(pkey);
+
+    /* Test that we can perform pkey operation */
+    EXPECT_SUCCESS(s2n_async_pkey_op_perform(pkey_op, pkey));
+
+    /* Test apply with different certificate chain only for sign operation */
+    if (pkey_op->type == S2N_ASYNC_SIGN) {
+        /* Create new chain and key, and modify current server conn */
+        struct s2n_cert_chain_and_key *chain_and_key_2 = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key_2,
+                S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+
+        /* Change server conn cert data */
+        EXPECT_NOT_NULL(conn->handshake_params.our_chain_and_key);
+        conn->handshake_params.our_chain_and_key = chain_and_key_2;
+
+        /* Test that async sign operation will fail as signature was performed over different private key */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_async_pkey_op_apply(pkey_op, conn), S2N_ERR_VERIFY_SIGNATURE);
+
+        /* Set signature validation mode to S2N_ASYNC_PKEY_VALIDATION_SKIP and test that async sign apply will pass now*/
+        EXPECT_SUCCESS(s2n_config_set_async_pkey_validation_mode(conn->config, S2N_ASYNC_PKEY_VALIDATION_SKIP));
+        EXPECT_SUCCESS(s2n_async_pkey_op_apply(pkey_op, conn));
+
+        /* Set chain_and_key back to original value and free new chain_and_key */
+        conn->handshake_params.our_chain_and_key = chain_and_key;
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_free(chain_and_key_2));
+
+        /* Increment flag counter for sign operation */
+        async_handler_called++;
+    } else {
+        /* Test decrypt operation passes */
+        EXPECT_SUCCESS(s2n_async_pkey_op_apply(pkey_op, conn));
+
+        /* Increment flag counter for decrypt operation */
+        async_handler_called++;
```

Comment:
You check in your test that async_handler_called==1. So if you increment async_handler_called for the decrypt operation, how do you know your signing tests are being run? If the handler is only called once, wouldn't that one call being a decrypt operation be an error?

Reply:
Path : Remove incrementing counter at decrypt operation. 

Question : 
Will it be good to not reset this value and increment it for sign operation and check if it incremented (check like value is greater than 0)? 
Reason : If i reset it to 0, then decrypt won't increase the counter.


QQ on current check :
Can decrypt (apply) operation pass without a sign (apply) operation?

Reply:
Done